src/codegen/codegen.cc

```diff
@@ -58,20 +60,122 @@ runtime::Module Build(const Array<LoweredFunc>& funcs,
   return m;
 }
 
+/*! \brief Helper class to serialize module */
+class ModuleSerializer {
+ public:
+  explicit ModuleSerializer(runtime::Module mod) : mod_(mod) {
+    Init();
+  }
+
+  void SerializeModule(dmlc::Stream* stream) {
+    // Only have one DSO module and it is in the root, then
+    // we will not produce import_tree_.
+    bool has_import_tree = true;
+    if (IsDSOModule(mod_->type_key()) && mod_->imports().empty()) {
+      has_import_tree = false;
+    }
+    uint64_t sz = 0;
+    if (has_import_tree) {
+      // we will append one key for _import_tree
+      // The layout is the same as before: binary_size, key, logic, key, logic...
+      sz = mod_vec_.size() + 1;
+    } else {
+      // Keep the old behaviour
+      sz = mod_->imports().size();
+    }
+    stream->Write(sz);
+
+    for (auto m : mod_vec_) {
+      std::string mod_type_key = m->type_key();
+      if (!IsDSOModule(mod_type_key)) {
+        stream->Write(mod_type_key);
+        m->SaveToBinary(stream);
+      } else {
+        if (has_import_tree) {
+          mod_type_key = "_lib";
+          stream->Write(mod_type_key);
+        }
+      }
+    }
+
+    // Write _import_tree key if we have
+    if (has_import_tree) {
+      std::string import_key = "_import_tree";
+      stream->Write(import_key);
+      stream->Write(import_tree_);
+    }
+  }
+
+ private:
+  void Init() {
+    CreateModuleIndex();
+    CreateImportTree();
+  }
+
+  void CreateModuleIndex() {
+    std::unordered_set<const runtime::ModuleNode*> visited {mod_.operator->()};
+    std::vector<runtime::ModuleNode*> stack {mod_.operator->()};
+    uint64_t module_index = 0;
+
+    while (!stack.empty()) {
+      runtime::ModuleNode* n = stack.back();
+      stack.pop_back();
+      mod2index_[n] = module_index++;
+      mod_vec_.emplace_back(n);
+      for (runtime::Module m : n->imports()) {
+        runtime::ModuleNode* next = m.operator->();
+        if (visited.count(next)) {
+          continue;
+        }
+        visited.insert(next);
+        stack.push_back(next);
+      }
+    }
+    import_tree_.resize(mod_vec_.size());
+  }
+
+  void CreateImportTree() {
+    std::vector<uint64_t> csr_row_ptr {0};
+    std::vector<uint64_t> csr_col_indices;
```

Comment:
directly store and serialize csr_row_ptr and csr_values. We can change the name to import_tree_row_ptr_, import_tree_child_indices_(values)